src/main/java/io/reactivex/internal/operators/parallel/ParallelPeek.java

```diff
@@ -0,0 +1,216 @@
+import io.reactivex.parallel.ParallelFlowable;
+import io.reactivex.plugins.RxJavaPlugins;
+
+/**
+ * Execute a Consumer in each 'rail' for the current element passing through.
+ *
+ * @param <T> the value type
+ */
+public final class ParallelPeek<T> extends ParallelFlowable<T> {
+
+    final ParallelFlowable<T> source;
+
+    final Consumer<? super T> onNext;
+    final Consumer<? super T> onAfterNext;
+    final Consumer<? super Throwable> onError;
+    final Action onComplete;
+    final Action onAfterTerminated;
+    final Consumer<? super Subscription> onSubscribe;
+    final LongConsumer onRequest;
+    final Action onCancel;
+
+    public ParallelPeek(ParallelFlowable<T> source,
+            Consumer<? super T> onNext,
+            Consumer<? super T> onAfterNext,
+            Consumer<? super Throwable> onError,
+            Action onComplete,
+            Action onAfterTerminated,
+            Consumer<? super Subscription> onSubscribe,
+            LongConsumer onRequest,
+            Action onCancel
+    ) {
+        this.source = source;
+
+        this.onNext = ObjectHelper.requireNonNull(onNext, "onNext");
+        this.onAfterNext = ObjectHelper.requireNonNull(onAfterNext, "onAfterNext");
+        this.onError = ObjectHelper.requireNonNull(onError, "onError");
+        this.onComplete = ObjectHelper.requireNonNull(onComplete, "onComplete");
+        this.onAfterTerminated = ObjectHelper.requireNonNull(onAfterTerminated, "onAfterTerminated");
+        this.onSubscribe = ObjectHelper.requireNonNull(onSubscribe, "onSubscribe");
+        this.onRequest = ObjectHelper.requireNonNull(onRequest, "onRequest");
+        this.onCancel = ObjectHelper.requireNonNull(onCancel, "onCancel");
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T>[] subscribers) {
+        if (!validate(subscribers)) {
+            return;
+        }
+
+        int n = subscribers.length;
+        @SuppressWarnings("unchecked")
+        Subscriber<? super T>[] parents = new Subscriber[n];
+
+        for (int i = 0; i < n; i++) {
+            parents[i] = new ParallelPeekSubscriber<T>(subscribers[i], this);
+        }
+
+        source.subscribe(parents);
+    }
+
+    @Override
+    public int parallelism() {
+        return source.parallelism();
+    }
+
+    static final class ParallelPeekSubscriber<T> implements Subscriber<T>, Subscription {
+
+        final Subscriber<? super T> actual;
+
+        final ParallelPeek<T> parent;
+
+        Subscription s;
+
+        boolean done;
+
+        ParallelPeekSubscriber(Subscriber<? super T> actual, ParallelPeek<T> parent) {
+            this.actual = actual;
+            this.parent = parent;
+        }
+
+        @Override
+        public void request(long n) {
+            try {
+                parent.onRequest.accept(n);
+            } catch (Throwable ex) {
+                Exceptions.throwIfFatal(ex);
+                RxJavaPlugins.onError(ex);
+            }
```

Comment:
`FlowableDoOnLifecycle` doesn't return either. There is no good way to report an error and not inject a lot of overhead. See `strict()`.